pkg/network/status/status.go

```diff
@@ -184,73 +185,69 @@ func (m *Prober) IsReady(ctx context.Context, ing *v1alpha1.Ingress) (bool, erro
 		cancel:       cancel,
 	}
 
-	var workItems []*workItem
+	workItems := make(map[string][]*workItem)
 
 	allProbeTargets, err := m.targetLister.ListProbeTargets(ctx, ing)
 	if err != nil {
 		return false, err
 	}
 
-	allPodIPs := sets.NewString()
+	// First, group the targets by IPs.
 	for _, target := range allProbeTargets {
-		// Iterate through sorted Pod IPs, for a consistent order.
-		for _, ip := range target.PodIPs.List() {
-			allPodIPs.Insert(ip)
-			// Each Pod is probed using the different hosts, protocol and ports until
-			// one of the probing calls succeeds. Then, the Pod is considered ready and all pending work items
-			// scheduled for that pod are cancelled.
-			ctx, cancel := context.WithCancel(ingCtx)
-			podState := &podState{
-				successCount: 0,
-				context:      ctx,
-				cancel:       cancel,
-			}
-
-			// Save the podState to be able to cancel it in case of Pod deletion
-			func() {
-				m.mu.Lock()
-				defer m.mu.Unlock()
-				m.podStates[ip] = podState
-			}()
-
-			// Update states and cleanup m.podStates when probing is done or cancelled
-			go func(ip string) {
-				<-podState.context.Done()
-				m.updateStates(ingressState, podState)
-
-				m.mu.Lock()
-				defer m.mu.Unlock()
-				// It is critical to check that the current podState is also the one stored in the map
-				// before deleting it because it could have been replaced if a new version of the ingress
-				// has started being probed.
-				if state, ok := m.podStates[ip]; ok && state == podState {
-					delete(m.podStates, ip)
-				}
-			}(ip)
-
+		for ip := range target.PodIPs {
 			for _, url := range target.URLs {
-				workItem := &workItem{
+				workItems[ip] = append(workItems[ip], &workItem{
 					ingressState: ingressState,
-					podState:     podState,
 					url:          url,
 					podIP:        ip,
 					podPort:      target.Port,
-				}
-				workItems = append(workItems, workItem)
+				})
+			}
+		}
+	}
+	for ip, ipWorkItems := range workItems {
+		ctx, cancel := context.WithCancel(ingCtx)
+		podState := &podState{
+			successCount: 0,
+			context:      ctx,
+			cancel:       cancel,
+		}
+
+		// Save the podState to be able to cancel it in case of Pod deletion
+		func() {
+			m.mu.Lock()
+			defer m.mu.Unlock()
+			m.podStates[ip] = podState
+		}()
+
+		// Update states and cleanup m.podStates when probing is done or cancelled
+		go func(ip string) {
+			<-podState.context.Done()
+			m.updateStates(ingressState, podState)
+
+			m.mu.Lock()
+			defer m.mu.Unlock()
+			// It is critical to check that the current podState is also the one stored in the map
+			// before deleting it because it could have been replaced if a new version of the ingress
+			// has started being probed.
+			if state, ok := m.podStates[ip]; ok && state == podState {
+				delete(m.podStates, ip)
 			}
+		}(ip)
```

Comment:
I think that the key bit here is that we now do this once per IP instead of potentially reseting things as we see each IP again in cases with multiple probe targets with overlapping ips (e.g. HTTP and HTTPS).

If we see the gateway scale, then only one of the probes associated with the PodIP is cancelled, and the other runs away forever because the podState isn't shared.

If I'm correct, then this fix is progress, but incomplete.  If I deploy two services simultaneously to the same gateway, then we may actually have the same problem because the `IsReady` for the first will initialize `podState` only to have it replaced by the other.  Then if cancellation is needed, only the surviving `podState` will be cancelled and the first service will probe forever.

Reply:
Yes, I will follow up with fixing by addressing https://github.com/knative/serving/issues/6269 .

Reply:
That's a separate issue than what I am describing.  Happy to jump on a call, if it's quicker.

Reply:
In the original version: https://github.com/knative/serving/blob/83300fb660d2d85b9c67e677e553da3978f07143/pkg/reconciler/ingress/status.go#L306 orphaned states are cancelled. There is no probing forever.
But this can be made cleaner by having a single context/channel per Pod IP to cancel all probing on that pod at once cleanly.

Reply:
I suspect that's what we want

Reply:
We still have `expireOldStates` https://github.com/knative/serving/pull/6406/files/0dd5f5d41df871eb68b86f222dbc50bb3af6dbb0#diff-4bc92f751e7fa71c25f068c120f3a20aL307 

For the cleaner fix that @JRBANCEL suggests we can follow up in a separate change

Reply:
So long as we follow up, I'm fine with forward progress and re-enabling the tests.  Have you looked at whether this fixes the IngressTLS test?